HTMLCanvasElement is orphaned causing a HTMLDocument leak on YouTube video pages

rreno opened the pull request:
#### ade36a336115d4a570eba730ef89fbfb168d3d21
<pre>
HTMLCanvasElement is orphaned causing a HTMLDocument leak on YouTube video pages
<a href="https://bugs.webkit.org/show_bug.cgi?id=256262">https://bugs.webkit.org/show_bug.cgi?id=256262</a>
rdar://108845985

Reviewed by Darin Adler.

CanvasRenderingContext2DBase has a State stack for setting up drawing
state. Two of the fields - strokeStyle and fillStyle - each hold CanvasStyle
objects. These objects may be set with colors or images or gradients. In
the case of a CanvasGradient, the gradient has a strong reference
to the context which creates a reference cycle.

CanvasRenderingContext2DBase -&gt; State -&gt; CanvasStyle -&gt; CanvasGradient -&gt; CanvasRenderingContext(2DBase)

What makes this cycle dangerous is the CanvasRenderingContext will
increase the ref count of an HTMLCanvasElement which, being a Node, will
increment the referencingNodeCount of a Document object. So if a
gradient is set on the context&apos;s fillStyle or strokeStyle we can cause a
Document leak if the state is never cleared like on YouTube video pages.

This patch changes the CanvasGradient object to hold a weak reference to
the CanvasRenderingContext instead of a strong reference which breaks the cycle.

* LayoutTests/fast/canvas
    Add tests to verify the now-decoupled lifetimes of gradient and
    context do not cause crashes.
* LayoutTests/fast/canvas/canvas-gradient-can-outlive-context-expected.txt: Added.
* LayoutTests/fast/canvas/canvas-gradient-can-outlive-context.html: Added.
* LayoutTests/fast/canvas/canvas-state-stack-gradient-expected.txt: Added.
* LayoutTests/fast/canvas/canvas-state-stack-gradient.html: Added.

* LayoutTests/http/tests/canvas
    Add a test to verify that setting the style of a 2D canvas context
    does not leak Documents.
* LayoutTests/http/tests/canvas/ctx.2d-canvas-style-no-document-leak-expected.txt: Added.
* LayoutTests/http/tests/canvas/ctx.2d-canvas-style-no-document-leak.html: Added.
* LayoutTests/http/tests/canvas/resources/background.png: Added.
* LayoutTests/http/tests/canvas/resources/ctx.2d-fillStyle-color.html: Added.
* LayoutTests/http/tests/canvas/resources/ctx.2d-fillStyle-gradient.html: Added.
* LayoutTests/http/tests/canvas/resources/ctx.2d-fillStyle-pattern.html: Added.

* Source/WebCore/html/canvas/CanvasGradient.cpp:
(WebCore::CanvasGradient::addColorStop):
* Source/WebCore/html/canvas/CanvasGradient.h:
* Source/WebCore/html/canvas/CanvasStyle.cpp:
(WebCore::parseColor):
    Since the gradient can outlive the context that created it, we need
    to be able to parse colors without having a CSSParserContext
    reference. This overload calls CSSParser::parseColorWithoutContext
    in that case.
* Source/WebCore/html/canvas/CanvasStyle.h:
(WebCore::CanvasStyle::canvasGradient const):

Canonical link: <a href="https://commits.webkit.org/263774@main">https://commits.webkit.org/263774@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/33294a1d7c6bd60985784081aa05bf8781cf3b63

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5641 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5794 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5984 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7193 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5627 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5640 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6018 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5767 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/7630 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5745 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5794 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5050 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7266 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3271 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5073 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/12630 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5140 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5150 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7029 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5589 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4574 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5038 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/5004 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1347 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9149 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5399 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/html/canvas/CanvasStyle.cpp:
```diff
@@ -159,8 +159,10 @@ void CanvasStyle::applyStrokeColor(GraphicsContext& context) const
+                return context.setStrokeGradient(gradient->gradient());
+            ASSERT_NOT_REACHED();
         },
         [&context] (const RefPtr<CanvasPattern>& pattern) {
```
Based on your diagnosis, I feel like we should be very suspicious of this `RefPtr<CanvasPattern>`, since that's just another kind of Gradient thing that might be created someplace and create another retain cycle. But perhaps that would make a great second patch.

rreno replied:
I checked and the CanvasPattern does not contain any references to the context. If that changes in the future I expect the test I added to catch that.

brentfulgham reviewed (commented):
I think this patch looks good, but we should get @grorg or @shallawa or even @darinadler to give it a quick look, too.

darinadler commented on Source/WebCore/html/canvas/CanvasGradient.h:
```diff
@@ -34,7 +34,7 @@ namespace WebCore {
 class Gradient;
 
-class CanvasGradient : public RefCounted<CanvasGradient> {
+class CanvasGradient : public RefCounted<CanvasGradient>, public CanMakeWeakPtr<CanvasGradient> {
```
Seems to me that it’s `CanvasGradient::m_context` which should be the `WeakPtr`.

rreno replied:
I wasn't sure what to do in `CanvasGradient::setColorStop` if the context had been destroyed. The standard doesn't give me an exception to throw and I wasn't sure if "do nothing" was the right answer.

darinadler replied:
The weak reference in the stack also isn’t correct; what keeps it alive? Nothing, which means the stack won’t work and we will fail to restore the gradient.

We should construct the actual test cases here for when one object outlives the other instead of doing thought experiments about the lifetimes.

rreno replied:
You're right; the test I wrote in this PR only verifies the Document isn't leaked it doesn't make any checks for the relative lifetimes of the objects I changed.

I wrote a test which crashes when the gradient has gone out of scope but the context is still alive so I agree now this approach doesn't work.

I'll try to write a test verifying that the context will live at least as long as the gradient and make the change you suggested.

rreno replied:
The test was more or less this:

```Javascript
if (true) { // make a scope
    let gradient = context.createLinearGradient(...);
    context.fillStyle = gradient
    // draw something
    context.save();
} // gradient goes out of scope

context.fillStyle = "green";
// draw something else
// try to force GC

context.restore();
if (typeof context.fillStyle == "string") // crash
    console.log("FAIL");

```

rreno replied:
I can pretty easily create a situation where the gradient outlives the context:

```Javascript
var gradient;
onload = () => {
    let context = document.createElement('canvas').getContext('2d');
    gradient = context.createLinearGradient(0, 0, 100, 100);
    gradient.setColorStop(0, 'red');
    ctx.fillStyle = gradient;
    // do some drawing
    requestAnimationFrame(() => forceGC()); // forceGC will do some work to make a GC happen then call finish below
};

function finish() {
    gradient.setColorStop(...); // null context here. Crashes on an ASSERT I added.
}
```

We end up collecting the context and the canvas element in `forceGC` but the global gradient is still alive. 

It looks like the only reason we even need a reference to the context is so we can call the CSS parser to parse colors with a context. And if I'm reading that code correctly, we only need the context to check for quirks. We can still do the fast path for color parsing without the document as a context, I think. Can we simply remove the context entirely from `CanvasGradient` and parse the color string without a CSSParserContext in `CanvasGradient::addColorStop`?

I'm not a CSS parser expert I defer to you. You're in the git blame for all of the parsing functions that we will call from here.

rreno replied:
If you're reading the above comment from your email, I apologize. I made some edits after submitting.

rreno replied:
Oh no that won't work for `OffscreenCanvas`. We still need to know whether to call the worker safe parser.

rreno replied:
I guess we could call either the parser the way we do now (when we have a context) or if the context has gone out of scope we fall back to `CSSParser::parseColorWithoutContext`. How does that sound?

Edit: Definitely not ideal  https://bugs.webkit.org/show_bug.cgi?id=235269

rreno replied:
Ok I'm going to write the code to do this and see how the tests turn out:

If we have a context, do what we do now
If the context is null, try to parse the color using `CSSParser::parseColorWithoutContext`
If that fails, throw the SyntaxError exception for an invalid color (even though it is possibly valid we just can't parse it at this time)

Alternatively we could just give up on parsing colors altogether when the context is null but then you can't mutate gradients across contexts and scopes which would probably be an observable regression (though that's unproven).

darinadler replied:
Your solution is OK but there is no reason we have to “parse wrong”. We can store a weak reference to the object that is actually providing the parsing context. Like the document itself. Just need to be sure we can do that for both onscreen and offscreen canvases. There is no reason that needs to be a pointer to the rendering context; in the case of an onscreen canvas I think it can be a weak pointer to the document. Figuring out the exact programming idiom to use won’t be trivial but it should not be terribly complicated.

darinadler reviewed (approved):
As I said I think we can do even better but this is OK.

rreno committed:
HTMLCanvasElement is orphaned causing a HTMLDocument leak on YouTube video pages
https://bugs.webkit.org/show_bug.cgi?id=256262
rdar://108845985

Reviewed by Darin Adler.

CanvasRenderingContext2DBase has a State stack for setting up drawing
state. Two of the fields - strokeStyle and fillStyle - each hold CanvasStyle
objects. These objects may be set with colors or images or gradients. In
the case of a CanvasGradient, the gradient has a strong reference
to the context which creates a reference cycle.

CanvasRenderingContext2DBase -> State -> CanvasStyle -> CanvasGradient -> CanvasRenderingContext(2DBase)

What makes this cycle dangerous is the CanvasRenderingContext will
increase the ref count of an HTMLCanvasElement which, being a Node, will
increment the referencingNodeCount of a Document object. So if a
gradient is set on the context's fillStyle or strokeStyle we can cause a
Document leak if the state is never cleared like on YouTube video pages.

This patch changes the CanvasGradient object to hold a weak reference to
the CanvasRenderingContext instead of a strong reference which breaks the cycle.

* LayoutTests/fast/canvas
    Add tests to verify the now-decoupled lifetimes of gradient and
    context do not cause crashes.
* LayoutTests/fast/canvas/canvas-gradient-can-outlive-context-expected.txt: Added.
* LayoutTests/fast/canvas/canvas-gradient-can-outlive-context.html: Added.
* LayoutTests/fast/canvas/canvas-state-stack-gradient-expected.txt: Added.
* LayoutTests/fast/canvas/canvas-state-stack-gradient.html: Added.

* LayoutTests/http/tests/canvas
    Add a test to verify that setting the style of a 2D canvas context
    does not leak Documents.
* LayoutTests/http/tests/canvas/ctx.2d-canvas-style-no-document-leak-expected.txt: Added.
* LayoutTests/http/tests/canvas/ctx.2d-canvas-style-no-document-leak.html: Added.
* LayoutTests/http/tests/canvas/resources/background.png: Added.
* LayoutTests/http/tests/canvas/resources/ctx.2d-fillStyle-color.html: Added.
* LayoutTests/http/tests/canvas/resources/ctx.2d-fillStyle-gradient.html: Added.
* LayoutTests/http/tests/canvas/resources/ctx.2d-fillStyle-pattern.html: Added.

* Source/WebCore/html/canvas/CanvasGradient.cpp:
(WebCore::CanvasGradient::addColorStop):
* Source/WebCore/html/canvas/CanvasGradient.h:
* Source/WebCore/html/canvas/CanvasStyle.cpp:
(WebCore::parseColor):
    Since the gradient can outlive the context that created it, we need
    to be able to parse colors without having a CSSParserContext
    reference. This overload calls CSSParser::parseColorWithoutContext
    in that case.
* Source/WebCore/html/canvas/CanvasStyle.h:
(WebCore::CanvasStyle::canvasGradient const):

Canonical link: https://commits.webkit.org/263774@main
